### PR TITLE
DOP-1590 : process objects once and add temporization to deletion

### PR DESCRIPTION
**Pull request by tkerdoncuff**
Add a parameter to temporize file deletion on the S3 bucket
The crawler now also adds .processed files to the bucket to ensure that files are only processed once between the moment they appear, and the moment they are deleted.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -84,48 +88,90 @@ func main() {
...
+			}
 
-		wg.Wait()
-		close(cDone)
-	}()
+			wg.Wait()
+			close(cDone)
+		}()
+		<-cDone
+	}
```
Comment (antonincms) [inline main.go:112]: Is there a reason why you're processing all files and then cleaning all files ? 
It leads to a bit of code duplication.

File: main.go [hunk 2]
```diff
@@ -84,48 +88,90 @@ func main() {
...
-		return elems
+		return unprocessed, processed
 	}
 	for _, s := range elems.Contents {
 		if strings.HasSuffix(*s.Key, suffix) {
-			ret.Contents = append(ret.Contents, s)
+			csvFiles = append(csvFiles, s)
+		}
+		if strings.HasSuffix(*s.Key, processedFlagSuffix) {
+			processedElems = append(processedElems, strings.Replace(*s.Key, processedFlagSuffix, suffix, -1))
```
Comment (antonincms) [inline main.go:156]: This is equivalent to the following
```suggestion
			processedElems = append(processedElems, strings.ReplaceAll(*s.Key, processedFlagSuffix, suffix))
```

File: main.go [hunk 3]
```diff
@@ -183,7 +256,20 @@ func processObject(
 				Msg("Unable to delete object")
 			return err
 		}
-	}
 
+		markerFileName := strings.Replace(*o.Key, opts.Suffix, opts.ProcessedFlagSuffix, -1)
```
Comment (antonincms) [inline main.go:260]: Idem 
```suggestion
		markerFileName := strings.ReplaceAll(*o.Key, opts.Suffix, opts.ProcessedFlagSuffix)
```

File: main.go [hunk 4]
```diff
@@ -169,9 +215,36 @@ func processObject(
 		return err
 	}
 
-	// Delete object
-	if opts.CleanObjects {
-		_, err = s3Cli.DeleteObject(ctx, &s3.DeleteObjectInput{
+	// Add .processed file to S3 bucket to avoid writing the same file to influx twice.
+	markerFileName := strings.Replace(*o.Key, opts.Suffix, opts.ProcessedFlagSuffix, -1)
```
Comment (antonincms) [inline main.go:219]: Idem 
```suggestion
	markerFileName := strings.ReplaceAll(*o.Key, opts.Suffix, opts.ProcessedFlagSuffix)
```

File: main.go [hunk 5]
```diff
@@ -62,13 +64,15 @@ func main() {
 		log.Fatal().Err(err).Msg("Unable to list objects")
 	}
 
-	filtered_res := filterBucketContent(*res, opts.Suffix)
+	unproc_csvs, proc_csvs := filterBucketContent(*res, opts.Suffix, opts.ProcessedFlagSuffix)
```
Comment (antonincms) [inline main.go:67]: It is more common to use camelCase in go : 
```suggestion
	unprocCsvs, procCsvs := filterBucketContent(*res, opts.Suffix, opts.ProcessedFlagSuffix)
```

File: main.go [hunk 6]
```diff
@@ -84,48 +88,90 @@ func main() {
...
+	}
 
-	<-cDone
 	log.Info().
 		Dur("elapsed", time.Since(start)).
 		Msg("Processing ended !")
 }
 
-func filterBucketContent(elems s3.ListObjectsOutput, suffix string) (ret s3.ListObjectsOutput) {
+func filterBucketContent(elems s3.ListObjectsOutput, suffix string, processedFlagSuffix string) (unprocessed s3.ListObjectsOutput, processed s3.ListObjectsOutput) {
```
Comment (antonincms) [inline main.go:142]:
```suggestion
func filterBucketContent(elems s3.ListObjectsOutput, suffix string, processedFlagSuffix string) (unprocessed, processed s3.ListObjectsOutput) {
```

File: main.go (hunk 1 above)
Comment (tkerdoncuff) [reply main.go:112]: Well they are not the same list. I guess I could replace both the procCsvs and unprocCSVs by a single map with a boolean as a value, iterate that then call either processObject or cleanObject. Thoughts ?

Comment (antonincms) [reply main.go:112]: Or maybe you could have a helper function, such as 
```
func parallelApply(list s3.ListObjectsOutput, fn func(o types.Object))  {
	var wg sync.WaitGroup
	cDone := make(chan bool)
	wg.Add(len(list))
	for _, item := range list {
		o := item
		go func() {
			defer wg.Done()
			fn(o)
		}()
	}

	wg.Wait()
}
```

And call it like so : 
```
	if len(unprocCsvs.Contents) > 0 {
		parallelApply(unprocCsvs, func(o types.Object) error {
			if processObject(ctx, dwn, upl, influxWriter, o); err != nil {
				log.Error().Err(err).Msg("Processing error")
			}
		})
	}

	if opts.CleanObjects && len(procCsvs.Contents) > 0 {
		parallelApply(procCsvs, func(o types.Object) error {
			if cleanObject(ctx, s3Cli, o); err != nil {
				log.Error().Err(err).Msg("Cleaning error")
			}
		})
	}
```

File: main.go [hunk 7]
```diff
@@ -86,52 +86,20 @@ func main() {
...
-				go func() {
-					defer wg.Done()
-					if err := processObject(ctx, dwn, upl, influxWriter, o); err != nil {
-						log.Error().Err(err).
-							Msg("Processing error")
-					}
-				}()
+	if len(unprocCsvs.Contents) > 0 {
+		parallelApply(unprocCsvs, func(o types.Object) {
+			if processObject(ctx, dwn, upl, influxWriter, o); err != nil {
```
Comment (antonincms) [inline main.go]:
```suggestion
			if err := processObject(ctx, dwn, upl, influxWriter, o); err != nil {
```

File: main.go [hunk 8]
```diff
@@ -86,52 +86,20 @@ func main() {
...
-				go func() {
-					defer wgClean.Done()
-					if err := cleanObject(ctx, s3Cli, o); err != nil {
-						log.Error().Err(err).
-							Msg("Processing error")
-					}
-				}()
+	if opts.CleanObjects && len(procCsvs.Contents) > 0 {
+		parallelApply(procCsvs, func(o types.Object) {
+			if cleanObject(ctx, s3Cli, o); err != nil {
```
Comment (antonincms) [inline main.go]:
```suggestion
			if err := cleanObject(ctx, s3Cli, o); err != nil {
```

File: pkg/flags/flags.go [hunk 9]
```diff
@@ -161,7 +161,7 @@ type Options struct {
 	Suffix              string        `long:"suffix" description:"Filename suffix to limit files read on the bucket."`
 	ProcessedFlagSuffix string        `long:"processed-flag-suffix" description:"Filename suffix to mark csv files as processed on the bucket." default:"processed"`
 	CleanObjects        bool          `long:"clean-objects" description:"Whether to delete S3 objects after processing them."`
-	S3MaxFileAge        time.Duration `long:"s3-max-file-age" description:"When cleanup is activated, only trigger deletion if csv is at leas this old." default:"10m"`
+	MaxObjectAge        time.Duration `long:"max-object-age" description:"When cleanup is activated, only trigger deletion if csv is at leas this old." default:"10m"`
```
Comment (antonincms) [inline pkg/flags/flags.go:164]:
```suggestion
	MaxObjectAge        time.Duration `long:"max-object-age" description:"When cleanup is activated, only trigger deletion if csv is at least this old." default:"10m"`
```

**Review by antonincms — APPROVED** [review]

**Pull request merged**